<a href="https://colab.research.google.com/github/Ali-Hamza-developer/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why


I chose the **Content Refresh Prioritization** lane. The purpose of this project is to identify web pages that are most likely to require content updates based on historical search performance and page characteristics. I selected this lane because manually reviewing thousands of pages is time-consuming, while machine learning can help prioritize which pages should be reviewed first. This can improve efficiency and support better content management decisions.

## The Question: Decision, Action, Cost of a Wrong Call

**Research Question:**
Which web pages should be prioritized for content refresh based on historical search performance and page characteristics?

**Decision:**
Determine which pages should be reviewed first.

**Action:**
An SEO analyst reviews and updates the highest-priority pages.

**Cost of a Wrong Recommendation:**
If the recommendation is incorrect, important declining pages may be missed or time may be wasted updating pages that do not require changes.

**Why Machine Learning?**
The dataset contains measurable features such as impressions, click-through rate (CTR), average position, and content age. Machine learning can analyze these signals together and produce a more effective ranking than a simple hand-written rule.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [2]:
import os
import sys
import subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)
elif os.path.basename(os.getcwd()) == "notebooks":
    os.chdir("..")

print("Current working directory:", os.getcwd())

Current working directory: /content/flyrank-ml-internship-starter


In [3]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("Dataset Shape:", df.shape)

print("\nTrend Direction Distribution")
print(df["trend_direction"].value_counts())

print("\nAverage CTR:", round(df["ctr"].mean(), 4))

print("Average Word Count:", round(df["word_count"].mean(), 2))

Dataset Shape: (30000, 44)

Trend Direction Distribution
trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64

Average CTR: 0.5107
Average Word Count: 3107.76


## Quick Look at the Data (2–3 Real Numbers)

The starter dataset contains **30,000 pages** and **44 features**, providing a good foundation for analyzing content performance.

The distribution of page trends shows that **16,262 pages are classified as declining**, while **5,962 are stable**, **4,388 are growing**, **2,236 are new**, and **1,152 are flat**. This indicates that a large portion of the dataset contains pages that may require review.

The dataset has an **average click-through rate (CTR) of 0.5107** and an **average word count of 3,107.76 words**. These measurable characteristics can be combined with other page features to help prioritize which pages should be reviewed for content refresh.

These observations suggest that the dataset contains meaningful variation and is suitable for building a decision-support system for content refresh prioritization.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

## Careful Words: What I Can and Can't Claim

This project provides decision support for identifying pages that may require content updates. The findings are based on observed historical data and should be interpreted as measured and directional rather than causal. This project does not predict Google's ranking algorithm or guarantee future search performance. Instead, it helps prioritize pages for review based on measurable patterns in the available data.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.